beacon-chain/blockchain/service.go

```diff
@@ -316,33 +317,40 @@ func (s *Service) originRootFromSavedState(ctx context.Context) ([32]byte, error
 	return genesisBlkRoot, nil
 }
 
-// initializeHeadFromDB uses the finalized checkpoint and head block found in the database to set the current head.
+// initializeHeadFromDB uses the finalized checkpoint and head block root from forkchoice to set the current head.
 // Note that this may block until stategen replays blocks between the finalized and head blocks
 // if the head sync flag was specified and the gap between the finalized and head blocks is at least 128 epochs long.
-func (s *Service) initializeHeadFromDB(ctx context.Context, finalizedState state.BeaconState) error {
+func (s *Service) initializeHead(ctx context.Context, st state.BeaconState) error {
 	cp := s.FinalizedCheckpt()
-	fRoot := [32]byte(cp.Root)
-	finalizedRoot := s.ensureRootNotZeros(fRoot)
-
-	if finalizedState == nil || finalizedState.IsNil() {
+	fRoot := s.ensureRootNotZeros([32]byte(cp.Root))
+	if st == nil || st.IsNil() {
 		return errors.New("finalized state can't be nil")
 	}
 
-	finalizedBlock, err := s.getBlock(ctx, finalizedRoot)
+	s.cfg.ForkChoiceStore.RLock()
+	root := s.cfg.ForkChoiceStore.HighestReceivedBlockRoot()
+	s.cfg.ForkChoiceStore.RUnlock()
+	blk, err := s.cfg.BeaconDB.Block(ctx, root)
 	if err != nil {
-		return errors.Wrap(err, "could not get finalized block")
+		return errors.Wrap(err, "could not get head block")
 	}
-	if err := s.setHead(&head{
-		finalizedRoot,
-		finalizedBlock,
-		finalizedState,
-		finalizedBlock.Block().Slot(),
-		false,
-	}); err != nil {
-		return errors.Wrap(err, "could not set head")
+	if root != fRoot {
+		st, err = s.cfg.StateGen.StateByRoot(ctx, root)
+		if err != nil {
+			return errors.Wrap(err, "could not get head state")
+		}
 	}
-
-	return nil
+	log.WithFields(logrus.Fields{
+		"root": fmt.Sprintf("%#x", root),
+		"slot": blk.Block().Slot(),
+	}).Info("Initialized head block from DB")
+	return errors.Wrap(s.setHead(&head{
+		root,
+		blk,
+		st,
+		blk.Block().Slot(),
+		false,
+	}), "could not set head")
 }
```

Comment:
nitpick: IMO this style is distracting compared to the typical:
```
if err := s.setHead(&head{root, blk, st, blk.Block().Slot(), false}); err != nil {
    return errors.Wrap(err, "could not set head")
}
return nil
```

Reply:
nitpick: we should avoid using positional struct fields where possible.

nitpick: It's also odd that `setHead` takes a struct pointer, when the pointer seems like a hack to make the func signature smaller - everywhere this is called, the struct is constructed. Looking at `setHead`, I would suggest breaking these fields out into individual args.

Reply:
This is a copy paste from develop, not sure what you want me to do here. I'll merge all the other suggestion and will take whatever you want on this one.

Reply:
For now suggestion is just to replace the highlighted lines with:
```
if err := s.setHead(&head{root, blk, st, blk.Block().Slot(), false}); err != nil {
    return errors.Wrap(err, "could not set head")
}
return nil
```

I'll send a PR later for how I'd like to change `setHead`.